File: junior/CE2002-A/Assignment/Code/HW07/QuickSort.java
class QuickSort {

    public QuickSort() {}

    public int[] Sort(int[] A, int lb, int rb) {
        if (lb >= rb) {
            return A;
        }
    
        int pivot = A[rb];
        int l = lb;
        int r = rb - 1;
    
        while (true) {
            while (A[l] < pivot) {
                l += 1;
            }
            while (A[r] >= pivot && r > lb) {
                r -= 1;
            }
            if (l < r) {
                int temp = A[l];
                A[l] = A[r];
                A[r] = temp;
                for (int j = 0; j < A.length; j++) {
                    if (j == A.length - 1) {
                        System.out.println(A[j]);
                    } else {
                        System.out.print(A[j]);
                        System.out.print(" ");
                    }
                }
            } else {
                break;
            }
        }
    
        if (A[rb] != A[l]) {
            int temp = A[rb];
            A[rb] = A[l];
            A[l] = temp;
            for (int j = 0; j < A.length; j++) {
                if (j == A.length - 1) {
                    System.out.println(A[j]);
                } else {
                    System.out.print(A[j]);
                    System.out.print(" ");
                }
            }
        }

        Sort(A, lb, l - 1);
        Sort(A, l + 1, rb);
        return A;
    }
}

File: ce/ce2002-a/Assignment/Code/HW07/quick_sort.py
def quick_sort(arr, lb, rb) :
    
    if (lb >= rb) :
        return arr

    pivot = arr[rb]
    l = lb
    r = rb - 1

    while True :
        while (arr[l] < pivot) :
            l += 1
        
        while (arr[r] >= pivot and r > lb) :
            r -= 1

        if (l < r) :
            temp = arr[l]
            arr[l] = arr[r]
            arr[r] = temp
        else :
            break

    if (arr[rb] != arr[l]) :
        temp = arr[rb]
        arr[rb] = arr[l]
        arr[l] = temp

    quick_sort(arr, lb, l - 1)
    quick_sort(arr, l + 1, rb)
    
    return arr